### PR TITLE
refactor: use tracker-pattern for exporting ProgramNotificationTemplate 

### PR DESCRIPTION
**Pull request by zubaira**
**Description:** This PR primarily focuses on refactoring the `ProgramNotificationTemplate` model. It aims to adopt the same export pattern that we have used for other tracker entities for example `Enrollment`.

**Changes:**
-   **Controller Updates:**
      - The GET method now uses `ProgramNotificationTemplateRequestParams` for user input.
      - A mapper is added to validate and map user input to `ProgramNotificationTemplateOperationParams`.

-   **Service layer validation and mapping:**
    - Created a mapper to validate and map `ProgramNotificationTemplateOperationParams` to `ProgramNotificationTemplateQueryParams` for use in the repository.
-   **Tests:**
	- Added new test in `ProgramNotificationTemplateControllerTest`
	- Created unit tests for both mapper classes.

### CODE REVIEW

**Review by muilpp — APPROVED** [review]

**Comment by muilpp** [comment]
If we want to adopt the same export pattern that we already use in `trackedEntities`, `enrollments`, and `events`, and you are the one doing the refactoring, should the program notification feature live under the tracker module?

**Comment by zubaira** [comment]
> @zubaira If we want to adopt the same export pattern that we already use in `trackedEntities`, `enrollments`, and `events`, and you are the one doing the refactoring, should the program notification feature live under the tracker module?

@muilpp 
While it may seem logical to place the notification feature within the tracker module due to its relationship with enrollments, and events, it might not be the best approach. Notifications often serve a broader purpose and can extend beyond the specific scope of the tracker module for example DataSetNotificationTemplate. 
The rationale for adopting this pattern was to standardize validation and error messages across all tracker-related modules.

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/notification/ProgramNotificationTemplateOperationParamsMapperTest.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.mockito.junit.jupiter.MockitoSettings;
+import org.mockito.quality.Strictness;
+
+@MockitoSettings(strictness = Strictness.LENIENT)
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/notification/ProgramNotificationTemplateOperationParamsMapperTest.java:54]: Could we not use LENIENT?
This test class has no so many cases so it is better to repeat the mocks inside each tests where needed

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/program/notification/ProgramNotificationTemplateServiceTest.java [hunk 2]
```diff
@@ -79,11 +81,11 @@ void setUp() {
     organisationUnitService.addOrganisationUnit(organisationUnit);
     program = createProgram('P');
     program.setAutoFields();
-    program.setUid("P_UID_1");
+    program.setUid(CodeGenerator.generateUid());
```
Comment (enricocolasante) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/program/notification/ProgramNotificationTemplateServiceTest.java:84]: If you don't need to set the UID you can remove this line.
`createProgram` already assign a UID to it.

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/program/notification/ProgramNotificationTemplateServiceTest.java [hunk 3]
```diff
@@ -79,11 +81,11 @@ void setUp() {
...
     programService.addProgram(program);
     programStage = createProgramStage('S', program);
     programStage.setAutoFields();
-    programStage.setUid("PS_UID_1");
+    programStage.setUid(CodeGenerator.generateUid());
```
Comment (enricocolasante) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/program/notification/ProgramNotificationTemplateServiceTest.java:88]: same here

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/notification/ProgramNotificationTemplateRequestParamsMapper.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                    requestParams.getPageSize(), NotificationPagingParam.DEFAULT_PAGE_SIZE)
+                : null)
+        .build();
+  }
+
+  private void validateRequestParams(ProgramNotificationTemplateRequestParams requestParams)
+      throws ConflictException {
+
+    if (requestParams.getProgram() == null && requestParams.getProgramStage() == null) {
+      throw new ConflictException("Program or ProgramStage must be specified.");
```
Comment (enricocolasante) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/notification/ProgramNotificationTemplateRequestParamsMapper.java:74]:
```suggestion
      throw new ConflictException("`program` or `programStage` must be specified.");
```

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/notification/ProgramNotificationTemplateRequestParamsMapper.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    if (requestParams.getProgram() != null && requestParams.getProgramStage() != null) {
+      throw new ConflictException("Program and ProgramStage cannot be processed together.");
```
Comment (enricocolasante) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/notification/ProgramNotificationTemplateRequestParamsMapper.java:78]:
```suggestion
      throw new ConflictException("`program` and `programStage` cannot be specified simultaneously.");
```

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/notification/ProgramNotificationTemplateRequestParamsMapper.java (hunk 4 above)
Comment (enricocolasante) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/notification/ProgramNotificationTemplateRequestParamsMapper.java:74]: We should refere to the parameters as the client is passing them, so it is `program` and not Program

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/notification/ProgramNotificationTemplateOperationParamsMapper.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  private <T extends BaseIdentifiableObject> T getEntity(UID objectId, Class<T> klass)
+      throws IllegalQueryException {
+    if (objectId == null) {
+      return null;
+    }
+
+    return Optional.ofNullable(manager.get(klass, objectId.getValue()))
+        .orElseThrow(
+            () ->
+                new IllegalQueryException(
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/notification/ProgramNotificationTemplateOperationParamsMapper.java:76]: Should this be a `NotFoundException` or a `BadRequestException`?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/notification/ProgramNotificationTemplateOperationParamsMapper.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+@RequiredArgsConstructor
+public class ProgramNotificationTemplateOperationParamsMapper {
+
+  private final IdentifiableObjectManager manager;
+
+  @Transactional
+  public ProgramNotificationTemplateQueryParams map(
+      ProgramNotificationTemplateOperationParams operationParams) {
+
+    Program program = getEntity(operationParams.getProgram(), Program.class);
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/notification/ProgramNotificationTemplateOperationParamsMapper.java:54]: I personally think that using `ProgramService` and `ProgramStageService` is more clear here but I do not have strong opinions on that

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/notification/ProgramNotificationTemplateOperationParamsMapper.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+      throws IllegalQueryException {
+    if (objectId == null) {
+      return null;
+    }
+
+    return Optional.ofNullable(manager.get(klass, objectId.getValue()))
+        .orElseThrow(
+            () ->
+                new IllegalQueryException(
+                    "%s with ID %s does not exist."
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/notification/ProgramNotificationTemplateOperationParamsMapper.java:77]:
```suggestion
                    "%s with UID %s does not exist."
```

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/notification/NotificationPagingParam.java [hunk 9]
```diff
@@ -46,5 +46,8 @@ public class BaseNotificationParam {
 
   private Integer pageSize;
 
+  @Deprecated(since = "2.41")
```
Comment (enricocolasante) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/notification/NotificationPagingParam.java:49]: This is not really a request param, right?
So it doesn't make sense to deprecate it

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/notification/NotificationPagingParam.java [hunk 10]
```diff
@@ -37,7 +37,7 @@
 @Data
 @AllArgsConstructor
 @NoArgsConstructor
-public class BaseNotificationParam {
+public class NotificationPagingParam {
```
Comment (enricocolasante) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/notification/NotificationPagingParam.java:40]: Is this class needed at all?

**Review by enricocolasante — CHANGES_REQUESTED** [review]
We are getting the pagination params but no pagination is actually implemented.
Is this something that will come later?

**Review by enricocolasante — COMMENTED** [review]
Just saw this one https://github.com/dhis2/dhis2-core/pull/18344 :)

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/notification/ProgramNotificationTemplateOperationParamsMapperTest.java (hunk 1 above)
Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/notification/ProgramNotificationTemplateOperationParamsMapperTest.java:54]: centralizing the mock setup in the @BeforeEach method enhances readability and reduces redundancy, especially when the same mocks are used across multiple test cases.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/notification/NotificationPagingParam.java (hunk 9 above)
Comment (zubaira) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/notification/NotificationPagingParam.java:49]: https://github.com/dhis2/dhis2-core/blob/b59d61119be81e63b681d64e474155bb242e7afc/dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/event/ProgramNotificationTemplateController.java#L86

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/notification/NotificationPagingParam.java (hunk 10 above)
Comment (zubaira) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/notification/NotificationPagingParam.java:40]: i think so. Providing paging information to `ProgramNotificationInstanceParam`, `ProgramNotificationTemplateOperationParams`

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/notification/ProgramNotificationTemplateOperationParamsMapperTest.java (hunk 1 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/notification/ProgramNotificationTemplateOperationParamsMapperTest.java:54]: Overusing `@MockitoSettings(strictness = Strictness.LENIENT)` is not enhancing the readability of the tests as all the setup for all the tests is placed in one method.
To make the tests more readable we should have anything that is common to all the tests in `@beforeEach` setup and anything that is not common, set in the test itself

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/notification/NotificationPagingParam.java (hunk 9 above)
Comment (enricocolasante) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/notification/NotificationPagingParam.java:49]: As I was saying, we should deprecate only fields that are actually exposed to the API.
This is an internal object created by some mapping, isn't it?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/notification/NotificationPagingParam.java (hunk 10 above)
Comment (enricocolasante) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/notification/NotificationPagingParam.java:40]: Maybe we can re-evaluate this after the pagination in store is completed as well

**Review by enricocolasante — APPROVED** [review]

**Pull request merged**